### ISSUE TITLE
Build Kivy iOS with arm and arm64 support

### ISSUE THREAD

**Issue by jeffsenn** [kivy/kivy-ios#115]
Apple will soon be requiring App Store submissions to include arm64 support:
https://developer.apple.com/news/?id=10202014a

Currently I think this is blocked by:  https://github.com/kivy/pyobjus/issues/17

This will also require changes to build scripts and a newer version of libffi that supports arm64

**Comment by tito** [comment]
I started it, in a branch named poly-arch. Bash script are gone, and trying to make it right also for the simulator. It's hard.

I also have few questions in mind that i don't know how exactly it will go. As the compilation defines how Python behave for few things, but i wonder how libraries will act on multi-arch as well. (sometimes, a python file is generated, or .h, depending the architecture. Theses files are used by the python interpreter. When the case happen, which one do i copy on the target?)

Etc. Not easy.

**Comment by jeffsenn** [comment]
I started this as well. What do you mean by "bash script are gone"? My main issue is the liblink/biglink hack to get shared extensions to built as static libraries.  I needed a newer Python (or at least distutils) to support fat arm builds and to rework scripts to use lipo to deal with fat builds in several places. I'm wondering whether it would be easier to build single-arch libs separately and then glue them together...

**Comment by tito** [comment]
That's the point. There is no more build-XXX.sh, only recipes approach like python-for-android, but in Python. Build every platform + lipo at the end for each recipe.

**Comment by jeffsenn** [comment]
Got a chance to check this out.  Nice!  Though looks like latest check in breaks building multiple arch since "build_all" state flag and the "cache_execution" decorator key is shared between all arch calls to "build" (That is: once it builds the first arch, it assumes others are already done.)  Maybe you've already fixed it... hacking around that gets me a build.  What is the plan for fixing XCode project build stuff? Are you going to make toolchain.py methods, or patch up the shell scripts in tools/ ?

**Comment by tito** [comment]
thanks for reporting the cache_execution, it was a test to see if i can speedup the next run, and later, provide a "status" of what has already been built for later usage. (ie, if i do binary release, i can precompile everything and delete the build directory, only the toolchain / recipe / dist will remain. :))

Will fix this issue. About Xcode, that was my job today, but then i saw that i don't install hostpython, therefore, no real way to use hostpython during xcode compilation.

I'm still unsure of the final layout, but i want to remove shell scripts.

**Comment by jeffsenn** [comment]
Coming along great! Builds for me now (though I've tried no complex examples).  I noticed you'll need this (currently if you happen to have harfbuzz installed then freetype will add a dependency).  It's a one liner:

--- a/recipes/freetype/__init__.py
+++ b/recipes/freetype/__init__.py
@@ -21,6 +21,7 @@ class FreetypeRecipe(Recipe):
                 "LDFLAGS={}".format(build_env["LDFLAGS"]),
                 "--prefix=/",
                 "--host={}".format(arch.triple),
+                "--without-harfbuzz",                
                 "--without-png",
                 "--without-bzip2",
                 "--without-fsspec",

**Comment by tito** [comment]
@jeffsenn Thanks for your feedback, line merged. You can now create xcode project with "toolchain create" command. It uses cookiecutter, and Remolten on irc said yaml was needed.

**Comment by Remolten** [comment]
Hey Remolten here.  Like Tito said, I had to install yaml, jinja2, and binaryornot to get it working.  Everything seems to be working now.

**Comment by jeffsenn** [comment]
Yep. Confirmed - seems to build and run for me on simulator and device.  Great work, thanks!

**Comment by tito** [comment]
Didn't tested on device yet. @jeffsenn is the device is arm64? Cause that's the only thing i will not be able to test.

Appart from that, how do you feel the new toolchain approach? Are you interested in coding recipes? Or just using it?

**Comment by jeffsenn** [comment]
Hm.. I hadn't tested more than getting python to load up and run.  Went and borrowed an iPhone 6 and tried to run the Pong demo - sadly it runs but does not render (correctly? at all?).  Does run fine on my iPhone5.  So appears there is something somewhat subtly wrong.  Don't have much time to pursue right this moment.  The "mode" debug output below seems like it might be suspicious...

<pre>
2015-02-16 10:26:33.102 pongu[4486:1458878] Available orientation: KIVY_ORIENTATION=LandscapeLeft LandscapeRight Portrait PortraitUpsideDown
2015-02-16 10:26:33.103 pongu[4486:1458878] PythonHome is: /private/var/mobile/Containers/Bundle/Application/BC4AA237-987B-4A5B-A990-EB384600D866/pongu.app
2015-02-16 10:26:33.104 pongu[4486:1458878] Initializing python
2015-02-16 10:26:33.362 pongu[4486:1458878] Running main.pyo: /private/var/mobile/Containers/Bundle/Application/BC4AA237-987B-4A5B-A990-EB384600D866/pongu.app/YourApp/main.pyo
[INFO              ] [Kivy        ] v1.9.0-dev
[INFO              ] [Python      ] v2.7.1 (r271:86832, Feb 16 2015, 08:24:14) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.56)]
[INFO              ] [Factory     ] 173 symbols loaded
[INFO              ] [Image       ] Providers: img_imageio, img_tex (img_dds, img_sdl2, img_ffpyplayer, img_pil, img_gif ignored)
[INFO              ] [OSC         ] using <thread> for socket
[INFO              ] [Window      ] Provider: sdl2
('MODE', 1, 1576521724)
('DRAWABLE SIZE', 320, 568)
[INFO              ] [GL          ] OpenGL version <OpenGL ES 2.0 Apple A8 GPU - 50.6.11>
[INFO              ] [GL          ] OpenGL vendor <Apple Inc.>
[INFO              ] [GL          ] OpenGL renderer <Apple A8 GPU>
[INFO              ] [GL          ] OpenGL parsed version: 2, 0
[INFO              ] [GL          ] Shading version <OpenGL ES GLSL ES 1.00>
[INFO              ] [GL          ] Texture max size <4096>
[INFO              ] [GL          ] Texture max units <8>
[INFO              ] [Window      ] auto add sdl input provider
[INFO              ] [Window      ] virtual keyboard not allowed, single mode, not docked
[INFO              ] [Text        ] Provider: sdl2
[INFO              ] [Base        ] Start application main loop
[INFO              ] [GL          ] NPOT texture support is available
</pre>

**Comment by tito** [comment]
It is. I'll try to test in simulator again.

**Comment by tito** [comment]
At least the Drawable size is wrong, it should be much higher. That's might be because there is no Launch Image for the device? (i discovered last week that If you don't provide Launch Image for all the correct size, iOS wont report you the correct window size.)

**Comment by Remolten** [comment]
I got my app working on the simulator for all devices except the iPhone 6 Plus (which crashes because it can't find a window provider).  I haven't tested on a physical device yet but I probably will soon.

**Comment by jeffsenn** [comment]
Not sure about the size thing.  That seems right if it's in "points" (or whatever) rather than "pixels" -- ("320x568" works on my iPhone 5).  Though the 5s and after do some exotic re-sampling to be resolution independent...  I don't claim to know what is going on; I haven't looked at all into SDL/OpenGL on iOS to have the slightest clue whether those numbers represent actual pixels or not.

**Comment by jeffsenn** [comment]
Also- Yes, I like the (python) recipe approach.  I'd say I'm _sort of_ interested in writing a couple. I would like to get  "stackless" python (a variant of 2.7.8) working (as an alternative to stock).  Which I've started, but is difficult due to changes ("improvements"?) to configure/build and distutils in Python involving cross-compiling circa 2.7.5.  Every time I have to deal with autoconf/etc I get a headache... :-)

**Comment by jeffsenn** [comment]
@tito - what's with recipes/python/ctypes_duplicate_longdouble.patch ?  seems to be a big binary patch in it.... was that intentional?

**Comment by niavlys** [comment]
> Appart from that, how do you feel the new toolchain approach?

At a first glance, it's much more pythonic than the previous version, so it looks promising! ;)

> Are you interested in coding recipes? 

Well, to be honest, not specially... but there's certainly times when you have to roll up your sleeves and get to work, so I've start digging for the repices I'm using (pyobjus, openssl, ffmpeg mainly), by mimicking what you've already done; it's far from being functionnal...

any help/docs/guidelines would be appreciated!

**Comment by jeffsenn** [comment]
(Slight followup): I tested again on arm64 (my alternate python build).  Same results.  However I did notice that there is a narrow (1 pixel or possibly 1 point) line down the left side of the screen indicating that it possibly is trying to resample into a very narrow and tall viewport (or something like that).

**Comment by niavlys** [comment]
@tito : basic kivy's examples run fine even on arm64 devices, but when I tried to build an archive through xcode => product => archive , clang's linker raises an error due to bad path:

```
Ld /Users/sylvain/Library/Developer/Xcode/DerivedData/isatis-fgtlwernehldbwajlimxdpaimhkj/Build/Intermediates/ArchiveIntermediates/isatis/IntermediateBuildFilesPath/isatis.build/Release-iphoneos/isatis.build/Objects-normal/armv7/isatis normal armv7
    cd /Users/sylvain/MyKivyBuild/kivy-ios/isatis-ios
    export IPHONEOS_DEPLOYMENT_TARGET=8.1
    export PATH="/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/usr/bin:/Applications/Xcode.app/Contents/Developer/usr/bin:/Library/Frameworks/Python.framework/Versions/2.7/bin:/opt/android-ndk-r9d:/opt/android-sdk-macosx/platform-tools:/opt/android-sdk-macosx/tools:/opt/local/bin:/opt/local/sbin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/opt/X11/bin"
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -arch armv7 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS8.1.sdk -L/Users/sylvain/Library/Developer/Xcode/DerivedData/isatis-fgtlwernehldbwajlimxdpaimhkj/Build/Intermediates/ArchiveIntermediates/isatis/BuildProductsPath/Release-iphoneos -L/Users/sylvain/MyKivyBuild/kivy-ios/isatis-ios/../build/lib -L/Users/tito/code/ios/kivy-ios/dist/lib -L/Users/tito/code/ios/kivy-ios/dist/root/python/lib -L/Users/tito/code/ios/kivy-ios/dist/root/python/lib/python2.7/config -F/Users/sylvain/Library/Developer/Xcode/DerivedData/isatis-fgtlwernehldbwajlimxdpaimhkj/Build/Intermediates/ArchiveIntermediates/isatis/BuildProductsPath/Release-iphoneos -filelist /Users/sylvain/Library/Developer/Xcode/DerivedData/isatis-fgtlwernehldbwajlimxdpaimhkj/Build/Intermediates/ArchiveIntermediates/isatis/IntermediateBuildFilesPath/isatis.build/Release-iphoneos/isatis.build/Objects-normal/armv7/isatis.LinkFileList -all_load -fobjc-link-runtime -miphoneos-version-min=8.1 -lios -lffi -framework MobileCoreServices -lc++ -framework Accelerate -framework CoreMotion -lfreetype -framework ImageIO -lsdl2_image -lsdl2 -lsqlite3 -framework MessageUI -lbz2 -lz -lsdl2_ttf -lpython -lsdl2_mixer -framework Foundation -framework UIKit -framework CoreGraphics -lkivy -framework OpenGLES -framework AudioToolbox -framework QuartzCore -Xlinker -dependency_info -Xlinker /Users/sylvain/Library/Developer/Xcode/DerivedData/isatis-fgtlwernehldbwajlimxdpaimhkj/Build/Intermediates/ArchiveIntermediates/isatis/IntermediateBuildFilesPath/isatis.build/Release-iphoneos/isatis.build/Objects-normal/armv7/isatis_dependency_info.dat -o /Users/sylvain/Library/Developer/Xcode/DerivedData/isatis-fgtlwernehldbwajlimxdpaimhkj/Build/Intermediates/ArchiveIntermediates/isatis/IntermediateBuildFilesPath/isatis.build/Release-iphoneos/isatis.build/Objects-normal/armv7/isatis

ld: warning: directory not found for option '-L/Users/sylvain/MyKivyBuild/kivy-ios/isatis-ios/../build/lib'
ld: warning: directory not found for option '-L/Users/tito/code/ios/kivy-ios/dist/lib'
ld: warning: directory not found for option '-L/Users/tito/code/ios/kivy-ios/dist/root/python/lib'
ld: warning: directory not found for option '-L/Users/tito/code/ios/kivy-ios/dist/root/python/lib/python2.7/config'
ld: library not found for -lios
clang: error: linker command failed with exit code 1 (use -v to see invocation) 
```
I Haven't yet figured out where and how these path were generated.

**Comment by tito** [comment]
@jeffsenn i fixed the ctypes patch, one of my vim swap file was in the diff :/
@niavlys maybe it's possible that ios library was not included before, but it is now. I recently added the dependency for kivy. Could you try to rebuild and recreate the project?

Also, the latest version update (only add, never remove) all the frameworks/libraries of your xcodeproject depending of the compiled recipes, when you create a new xcodeproject (with create). I've added "update" if you want to update an existing project.

**Comment by niavlys** [comment]
> Could you try to rebuild and recreate the project?
I still have these warnings, but at least no error with clang's linker.

**Comment by tito** [comment]
News: hostpython supports ctypes (needed for setuptools-based setup.py). python recipe reduce its own installation (packing all the python libs into python27.zip). and now there is only 4 recipes left before switching to the new toolchain (even if it's not fully working, it's far better than before): plyer, pil, openssl, openssl-link.

**Comment by tito** [comment]
Would anybody available for a binary test? Ie, i'm compiling the whole toolchain, and you uses it (no more build)?

I'm seeking for an usage review of the toolchain. What's good, what's wrong, etc.

**Comment by jeffsenn** [comment]
would be nice to do this from a stock mac install (so as to note any prerequisite python modules dependencies or build tools). 

A "buildall" option would be useful :-)

Is there a good test project (one that has enough coverage) so we can bring it up and jump through some hoops to declare "success"?

Today: 

I notice I need "yasm" for ffmepg  -- (brew install yasm) of course comes to the rescue....

But then I wind up here (and I'm not sure... do I need to patch my XCode toolchain somehow? Merely having gas-preprocessor on my path seems not sufficient)

<pre>
Build ffmpeg for armv7
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
GNU assembler not found, install/update gas-preprocessor

If you think configure made a mistake, make sure you are using the latest
version from Git.  If the latest version fails, report the problem to the
ffmpeg-user@ffmpeg.org mailing list or IRC #ffmpeg on irc.freenode.net.
Include the log file "config.log" produced by configure as this will help
solve the problem.
</pre>

**Comment by tito** [comment]
ok, only gas-processor is required i think not yasm. Recipe need to be updated :/

**Comment by niavlys** [comment]
> Would anybody available for a binary test? Ie, i'm compiling the whole toolchain, and you uses it (no more build)?

could be really nice!
currently here some recipes don't work (pyobjus), other don't even build (ffmpeg, audiostream)...quite frustrating especially as we can see the hard job you're doing!

**Comment by jeffsenn** [comment]
audiostream needs something like this to build (though haven't tried it runtime yet):
<pre>
--- a/recipes/audiostream/__init__.py
+++ b/recipes/audiostream/__init__.py
@@ -1,4 +1,5 @@
 from toolchain import CythonRecipe
+from os.path import join
 
 
 class AudiostreamRecipe(CythonRecipe):
@@ -8,6 +9,10 @@ class AudiostreamRecipe(CythonRecipe):
     depends = ["python", "sdl2"]
     pre_build_ext = True
 
+    def get_recipe_env(self, arch):
+        env = super(AudiostreamRecipe, self).get_recipe_env(arch)
+        env["CC"] += " -I{}".format(join(self.ctx.dist_dir, "include", "common", "sdl2_mixer"))
+        return env
 
 recipe = AudiostreamRecipe()
 
</pre>

**Comment by tito** [comment]
pyobjus has been fixed a little bit before. Will retry to compile the whole toolchain from scratch. I'm using it for 2 customers app, it's a pure joy.

**Comment by tito** [comment]
@jeffsenn Thans for the patch. FYI, include path are added automatically depending the "depends", in the env["CFLAGS"]. Adding sdl2_mixer in the depends fix the issue. It's commited and pushed.

I'm going to push this new toolchain in the master branch.

**Comment by tito** [comment]
Pushed in master. Kivy documentation updated as well.
